### PR TITLE
Add custom consult code to rabbit_prelaunch_file module

### PR DESCRIPTION
**Pull request by lukebakken**
This will be used as part of the fix for rabbitmq/osiris#78

If a RabbitMQ `advanced.config` file contains the following:

```
{customize_hostname_check, [
    {match_fun, public_key:pkix_verify_hostname_match_fun(https)}
]}
```

...`file:consult/1` will fail because it does not evaluate terms in the
file.

The code in `rabbit_consult` was copied from this OTP module:

https://github.com/erlang/otp/blob/master/lib/ssl/src/ssl_dist_sup.erl

...and then modified for our use.

### CODE REVIEW

**Review by dumbbell — COMMENTED** [review]
To allow anyone to locate the exact module the code was copied from, could you please add a command stating the commit hash, and perhaps a link to GitHub to that exact version of the file?

File: deps/rabbit/apps/rabbitmq_prelaunch/src/rabbit_prelaunch_file.erl [hunk 1]
```diff
@@ -0,0 +1,83 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright Ericsson AB 2011-2022. All Rights Reserved.
+%% Copyright (c) 2022 VMware, Inc. or its affiliates.  All rights reserved.
+%%
```
[the remark this answers is not in the document]
Comment (lukebakken) [reply deps/rabbit/apps/rabbitmq_prelaunch/src/rabbit_prelaunch_file.erl]: I reverted the header to that used in the original.

**Review by michaelklishin — CHANGES_REQUESTED** [review]
A node that starts just fine with the following config in `master`:

``` erlang
[
  {rabbit, [
    {consumer_timeout, none}
  ]},

  {kernel, [
    {inet_dist_use_interface, {127,0,0,1}}
  ]}
].
```

no longer starts on this branch:

``` erlang
2022-06-01 19:10:39.247491+04:00 [debug] <0.138.0> Override with advanced configuration file "/path/to/RabbitMQ/configs/advanced.config"
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0> Failed to load advanced configuration file "/path/to/RabbitMQ/configs/advanced.config", incorrect format: [{rabbit,
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                                [{consumer_timeout,
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                                  none}]},
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                               {kernel,
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                                [{inet_dist_use_interface,
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                                  {127,
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                                   0,
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                                   0,
2022-06-01 19:10:39.250470+04:00 [error] <0.138.0>                                                                                                                                   1}}]}]
2022-06-01 19:10:39.252847+04:00 [error] <0.138.0>
2022-06-01 19:10:39.252847+04:00 [error] <0.138.0> BOOT FAILED
2022-06-01 19:10:39.252847+04:00 [error] <0.138.0> ===========
2022-06-01 19:10:39.252847+04:00 [error] <0.138.0> Error during startup: {error,failed_to_parse_advanced_configuration_file}
```

**Comment by michaelklishin** [comment]
Note that I used `RABBITMQ_CONFIG_FILE` and not `RABBITMQ_ADVANCED_CONFIG_FILE` to point to the file. Still, I'm afraid some users may still be using `rabbitmq.config` files in the Erlang term format to this day.

**Comment by lukebakken** [comment]
@michaelklishin thanks that is definitely a bug. I will investigate.

**Comment by lukebakken** [comment]
OK I have tested the latest changes by starting up a cluster using inter-node TLS *and* wildcard certs. Here's the repo with my configuration and certs:

https://github.com/lukebakken/rabbitmq-osiris-78

```
make FULL=1 HOSTNAME=shostakovich.local \
 RABBITMQ_CONF_ENV_FILE=/home/lbakken/development/lukebakken/rabbitmq-osiris-78/rabbitmq-env-local.conf \
 RABBITMQ_CONFIG_FILE=/home/lbakken/development/lukebakken/rabbitmq-osiris-78/rabbitmq-local.conf \
 RABBITMQ_ADVANCED_CONFIG_FILE=/home/lbakken/development/lukebakken/rabbitmq-osiris-78/advanced-local.config \
 PLUGINS='rabbitmq_management rabbitmq_top' \
 NODES=3 start-cluster
```

After starting the above, I can use my `ssl_client.py` program to connect to a node. TLS verification works as expected, even using wildcards.

To generate the wildcard certs, I did the following in `tls-gen/basic`:

```
make 'CN=*.local' gen-server
make 'CN=*.local' gen-client
```

I also added a host entry in `/etc/hosts` for `shostakovich.local` using the local IP address.

**Comment by lukebakken** [comment]
Docker and `docker compose` makes testing this easy:

```
https://github.com/lukebakken/rabbitmq-osiris-78
cd rabbitmq-osiris-78
make image
make up
pipenv install
pipenv run ssl_client
```

**Comment by lukebakken** [comment]
OK, the following converts `Fun`s before `jsx:encode` is called:

https://github.com/rabbitmq/rabbitmq-server/pull/4932/commits/1d3d8a92a5e24f94cae22584d2f3c8454199d49c

**Comment by lukebakken** [comment]
@michaelklishin ready for a re-review. When the OCI build finishes I'll re-test using this - https://github.com/rabbitmq/rabbitmq-server/pull/4932#issuecomment-1145326553

**Review by michaelklishin — APPROVED** [review]

**Comment by gomoripeti** [comment]
a leftover `file:format_error` is not applicable any more. Noticed by dialyzer. I think it will crash in case `rabbit_prelaunch_file:consult_file` returns an error.
https://github.com/rabbitmq/rabbitmq-server/pull/4932/files#diff-d41c11590653efa85e6caaa3cf071a5ecce15f3f118a705ed9cc393d657515bdR400

(sorry for not submitting a fix PR right away)

**Pull request merged**